A few finalizer thread safety fixes.

yuyichao opened the pull request:
Noticed the memset conditional issue when re-reading the code and then discovered a stupid mistake and a more subtle one during testing.....
- Fix wrong address to `memset` ...
- Clearing the memory in `finalize_object` should be unconditional and before the `cmpxchg`.
  
  Since even when the `cmpxchg` succeeds, the owning thread might have read
  the old length and set it to a different value based on the old length,
  leaving an uncleared hole.
- List length needs to be read again after acquiring the finalizer lock.
  
  Someone might have changed it and the resizing might not have the expected effect.

yuyichao committed:
A few finalizer thread safety fixes.

* Fix wrong address to `memset` ...

* Clearing the memory in `finalize_object` should be unconditional
  and before the `cmpxchg`.

  Since even when the `cmpxchg` succeeds, the owning thread might have read
  the old length and set it to a different value based on the old length,
  leaving an uncleared hole.

* List length needs to be read again after acquiring the finalizer lock.

  Someone might have changed it and the resizing might not have the expected effect.

vtjnash commented on src/gc.c:
```diff
@@ -249,6 +249,8 @@ static void gc_add_finalizer_(jl_ptls_t ptls, void *v, void *f)
     size_t oldlen = jl_atomic_load_acquire(&a->len);
     if (__unlikely(oldlen + 2 > a->max)) {
         JL_LOCK_NOGC(&finalizers_lock);
+        // `a->len` might have been modified.
```
doesn't that also imply we need to redo the `oldlen + 2 > a->max` test?

yuyichao replied:
It doesn't hurt to call `arraylist_grow` when the length is shrunk by another thread already (the grow will just not allocate any memory).

yuyichao replied:
(And it's less code this way since this is the slow path anyway...)

vtjnash commented on src/gc.c:
```diff
@@ -149,10 +149,10 @@ static void finalize_object(arraylist_t *list, jl_value_t *o,
+        jl_atomic_compare_exchange(&list->len, oldlen, len);
+        // The memset needs to be unconditional since the thread might have
+        // already read the length.
+        memset(&items[len], 0, (oldlen - len) * sizeof(void*));
```
afaict, we also need to avoid the `len -= 2` call above

yuyichao replied:
`len` is a local variable and it is only "published" (i.e. released) with the `cmpxchg` after all the content modifications are done so it is important that we update the `len` but the `memset` should be before the `cmpxchg` when the `cmpxchg` succeeds and is fine to be on either side if `cmpxchg` fails.

* For both `finalize_object` and `gc_add_finalizer_`, the synchronization happens through the `arraylist_t::len` field and only happens at the beginning and the end of the function.
* There can be at most two threads operating on the same list, one in `gc_add_finalizer_` and one in `finalize_object`
* `gc_add_finalizer_` only increase the length and mutate content after the length it reads whereas `finalize_object` only decrease the length and mutate the content before the length it reads, and all content mutation happens in between the acquire of the `len` and release (including the `seq_cst` ordering of the `cmpxchg`) of the `len`. This way, the mutation on a thread cannot be observed on another thread before the release of the `len` (and will be consistently observed after that).
* Finally it is important that the `len` increment always succeeds so `gc_add_finalizer_` uses a release store and the `finalize_object` use `cmpxchg`. The decrement doesn't have to succeed since it can clear the memory instead.

yuyichao commented:
Rebased with the `memset` order fixed. I also added more comment of the synchronization scheme used here. Comment on the scheme, comment (including grammar/spelling mistakes) welcome....

tkelman commented on src/gc.c:
```diff
@@ -119,7 +119,13 @@ static void run_finalizer(jl_ptls_t ptls, jl_value_t *o, jl_value_t *ff)
+    // to the first `oldlen` elements and no mutation is allowed after the
+    // new length is published with the `cmpxchg` at the end of the function.
+    // This way, the mutation should not conflict with the owning thread,
+    // which only write to after `len` and will not resize the buffer without
```
so the owning thread only writes to locations in the list later than `len`?

yuyichao replied:
Yes.

tkelman replied:
wasn't very clear that "after `len`" is referring to a position in the array from this wording

yuyichao commented:
The mac travis spends all it's time rebuilding openblas?

tkelman commented:
```
==> Installing staticfloat/julia/julia dependency: staticfloat/julia/openblas-julia
==> Downloading https://juliabottles.s3.amazonaws.com/openblas-julia-0.2.18.mavericks.bottle.tar.gz
/usr/bin/curl --fail --remote-time --location --user-agent Homebrew/0.9.9 (Macintosh; Intel Mac OS X 10.9.5) curl/7.30.0 https://juliabottles.s3.amazonaws.com/openblas-julia-0.2.18.mavericks.bottle.tar.gz -C 0 -o /Users/travis/Library/Caches/Homebrew/openblas-julia-0.2.18.mavericks.bottle.tar.gz.incomplete --silent
Error: Failed to download resource "openblas-julia"
Download failed: https://juliabottles.s3.amazonaws.com/openblas-julia-0.2.18.mavericks.bottle.tar.gz
Warning: Bottle installation failed: building from source.
```

yuyichao commented:
The test code is at https://github.com/yuyichao/explore/commit/fd44e68c74a053435bae7334c098bb9e16723c3b and the `memset` order seems to have fix a very low probability lost of finalizers that I've seen during the test....

tkelman commented on src/gc.c:
```diff
@@ -246,9 +254,20 @@ static void gc_add_finalizer_(jl_ptls_t ptls, void *v, void *f)
+    // synchronize with `finalize_object` on another thread. Apart from the GC,
+    // which is blocked by entering a unsafe region, there might be only
+    // one other thread accessing our list in `finalize_object`
+    // (only one thread since it need to acquire the finalizer lock).
```
needs to acquire

vtjnash commented:
lgtm,nice find. do you know any good resources on understanding load-acquire/store-release I could read?

yuyichao commented:
I mainly learned it from https://herbsutter.com/2013/02/11/atomic-weapons-the-c-memory-model-and-modern-hardware/ and https://www.youtube.com/watch?v=DS2m7T6NKZQ

tkelman commented on src/gc.c:
```diff
@@ -119,7 +119,13 @@ static void run_finalizer(jl_ptls_t ptls, jl_value_t *o, jl_value_t *ff)
+    // to the first `oldlen` elements and no mutation is allowed after the
+    // new length is published with the `cmpxchg` at the end of the function.
+    // This way, the mutation should not conflict with the owning thread,
+    // which only write to locations later than `len`
```
writes to

ViralBShah commented:
I assume we need this in 0.5, right?

yuyichao commented:
It's a bug fix so the feature freeze doesn't matter and I'm planning to merge this after the CI is ok.

Note that this should only happen in really rare conditions in real code (explicitly calling `finalize` with multithreads)
